pkg/drivers/ssh/ssh.go

```diff
@@ -0,0 +1,243 @@
+/*
+Copyright 2019 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ssh
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"path"
+	"strconv"
+	"time"
+
+	"github.com/docker/machine/libmachine/drivers"
+	"github.com/docker/machine/libmachine/engine"
+	"github.com/docker/machine/libmachine/log"
+	"github.com/docker/machine/libmachine/mcnutils"
+	"github.com/docker/machine/libmachine/state"
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2"
+	pkgdrivers "k8s.io/minikube/pkg/drivers"
+	"k8s.io/minikube/pkg/minikube/command"
+	"k8s.io/minikube/pkg/minikube/cruntime"
+	"k8s.io/minikube/pkg/minikube/sysinit"
+)
+
+// Driver is a driver designed to run kubeadm w/o VM management.
+// https://minikube.sigs.k8s.io/docs/reference/drivers/ssh/
+type Driver struct {
+	*drivers.BaseDriver
+	*pkgdrivers.CommonDriver
+	EnginePort int
+	SSHKey     string
+	runtime    cruntime.Manager
+	exec       command.Runner
+}
+
+// Config is configuration for the SSH driver
+type Config struct {
+	MachineName      string
+	StorePath        string
+	ContainerRuntime string
+}
+
+const (
+	defaultTimeout = 15 * time.Second
+)
+
+// NewDriver creates and returns a new instance of the driver
+func NewDriver(c Config) *Driver {
+	d := &Driver{
+		EnginePort: engine.DefaultPort,
+		BaseDriver: &drivers.BaseDriver{
+			MachineName: c.MachineName,
+			StorePath:   c.StorePath,
+		},
+	}
+	runner := command.NewSSHRunner(d)
+	runtime, err := cruntime.New(cruntime.Config{Type: c.ContainerRuntime, Runner: runner})
+	// Libraries shouldn't panic, but there is no way for drivers to return error :(
+	if err != nil {
+		klog.Fatalf("unable to create container runtime: %v", err)
+	}
+	d.runtime = runtime
+	d.exec = runner
+	return d
+}
+
+// DriverName returns the name of the driver
+func (d *Driver) DriverName() string {
+	return "ssh"
+}
+
+func (d *Driver) GetSSHHostname() (string, error) {
+	return d.GetIP()
+}
+
+func (d *Driver) GetSSHUsername() string {
+	return d.SSHUser
+}
+
+func (d *Driver) GetSSHKeyPath() string {
+	return d.SSHKeyPath
+}
+
+func (d *Driver) PreCreateCheck() error {
+	if d.SSHKey != "" {
+		if _, err := os.Stat(d.SSHKey); os.IsNotExist(err) {
+			return fmt.Errorf("SSH key does not exist: %q", d.SSHKey)
+		}
+
+		// TODO: validate the key is a valid key
```

Comment:
Either add issue number or remove TODO

Reply:
That was just copied from docker-machine